README
libopenjtalk
Dynamic library version of Open JTalk

document updated on 2011-02-15

based on Open JTalk
http://sourceforge.net/projects/open-jtalk/
see README.open_jtalk-1.01 and COPYING

libopenjtalk by Takuya Nishimoto
http://ja.nishimotz.com/project:libopenjtalk
(contribution of nishimotz is BSD lisence)
started on 2010-06-27

obsolated: Python Wrapper https://github.com/nishimotz/ojt-python
obsolated: branch nvdajp
obsolated: branch ubuntu_sjis (merged to master)
pending: branch ubuntu_utf16

=============================================
building libopenjtalk.dll for nvdajp 

requires: https://github.com/nishimotz/htsengineapi

use Windows XP 32bit, cygwin-1.7, mecab-0.98 for Win32, Python 2.7 Win32

use libopenjtalk master branch
use do_configure_mingw32.sh

$ export CXX='g++ -mno-cygwin'
$ export CC='gcc -mno-cygwin'
$ sh /usr/bin/set-gcc-default-3.sh

$ cd htsengineapi
$ autoconf
$ ./configure --build=i686-pc-mingw32
$ make
$ cd ..

$ cd libopenjtalk

$ sh do_configure_mingw32.sh
$ make
$ cd lib
$ make -f Makefile.mingw32 clean
$ make -f Makefile.mingw32

$ file libopenjtalk.dll
libopenjtalk.dll: PE32 executable for MS Windows (DLL) (console) Intel 80386 32-bit

$ strip libopenjtalk.dll

=============================================
using libopenjtalk.dll for nvdajp

see source/synthDrivers/_jtalk_core.py
http://bazaar.launchpad.net/~nishimotz/nvdajp/releases_2011.1

=============================================
testing libopenjtalk.dll for nvdajp

uses lib/jtalk.py

requires nvwave.py from nvdajp
requires libmecab.dll from http://mecab.sourceforge.net/

JT_DIR = r'C:\work\jtalk'
DIC = JT_DIR + r"\dic"     # shift_jis version of mecab dic is c:/work/jtalk/dic
VOICE = JT_DIR + r"\voice" # m001 voice is c:/work/jtalk/voice
MECAB_DLL = JT_DIR + r"\libmecab.dll"  # c:/work/jtalk/libmecab.dll
MECABRC = JT_DIR + r"\mecabrc"         # c:/work/jtalk/mecabrc
JT_DLL = JT_DIR + r"\libopenjtalk.dll" # c:/work/jtalk/libopenjtalk.dll

$ python jtalk.py

=============================================
usage for Ubuntu Linux (tested with 9.04)
(dic charset is Shift_JIS)

requires: https://github.com/nishimotz/htsengineapi

$ sudo apt-get install autoconf libtool
$ cd htsengineapi
$ autoreconf
$ autoconf
$ CC='gcc -fPIC' ./configure
$ make clean
$ make
$ cd ..
$ cd libopenjtalk
$ autoreconf
$ sh do_configure_ubuntu.sh
$ make clean
$ make
$ cd lib
$ make -f Makefile.ubuntu target

check symbols:
$ nm ./libs/libopenjtalk.so.0.0.0 | lv 

test (not working):
$ export LD_LIBRARY_PATH=.libs ; ./mytest

=============================================
building shift_jis dic for nvdajp 

see mecab-naist-jdic/
  eng_dic_maker.py
  tankan_dic_maker.py
  custom_dic_maker.py
  make_jdic.sh
requires:
http://cpansearch.perl.org/src/MASH/Lingua-JA-Yomi-0.01/lib/Lingua/JA/bep-eng.dic
nvdajp: source/nvdajp_dic.py



ChangeLog
2013-12-23 17:03  uratec

	* configure.ac, bin/Makefile.am: Modify audio compiler option for
	  audio devices.

2013-12-17 13:02  uratec

	* mecab/src/param.cpp: Add ctype.h for android-ndk.

2013-12-11 15:46  uratec

	* configure.ac: Update configure.ac for version 1.07.

2013-12-11 15:33  uratec

	* COPYING, NEWS, README, bin/open_jtalk.c, jpcommon/jpcommon.c,
	  jpcommon/jpcommon.h, jpcommon/jpcommon_label.c,
	  jpcommon/jpcommon_node.c,
	  jpcommon/jpcommon_rule_ascii_for_euc_jp.h,
	  jpcommon/jpcommon_rule_ascii_for_shift_jis.h,
	  jpcommon/jpcommon_rule_ascii_for_utf_8.h,
	  jpcommon/jpcommon_rule_euc_jp.h,
	  jpcommon/jpcommon_rule_shift_jis.h,
	  jpcommon/jpcommon_rule_utf_8.h, mecab/COPYING,
	  mecab/src/common.h, mecab/src/iconv_utils.cpp,
	  mecab/src/iconv_utils.h, mecab/src/libmecab.cpp,
	  mecab/src/mecab.cpp, mecab/src/mecab.h, mecab/src/param.cpp,
	  mecab/src/thread.h, mecab/src/utils.cpp, mecab/src/winmain.h,
	  mecab-naist-jdic/COPYING, mecab2njd/mecab2njd.c,
	  mecab2njd/mecab2njd.h, njd/njd.c, njd/njd.h, njd/njd_node.c,
	  njd2jpcommon/njd2jpcommon.c, njd2jpcommon/njd2jpcommon.h,
	  njd2jpcommon/njd2jpcommon_rule_ascii_for_euc_jp.h,
	  njd2jpcommon/njd2jpcommon_rule_ascii_for_shift_jis.h,
	  njd2jpcommon/njd2jpcommon_rule_ascii_for_utf_8.h,
	  njd2jpcommon/njd2jpcommon_rule_euc_jp.h,
	  njd2jpcommon/njd2jpcommon_rule_shift_jis.h,
	  njd2jpcommon/njd2jpcommon_rule_utf_8.h,
	  njd_set_accent_phrase/njd_set_accent_phrase.c,
	  njd_set_accent_phrase/njd_set_accent_phrase.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_ascii_for_euc_jp.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_ascii_for_shift_jis.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_ascii_for_utf_8.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_euc_jp.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_shift_jis.h,
	  njd_set_accent_phrase/njd_set_accent_phrase_rule_utf_8.h,
	  njd_set_accent_type/njd_set_accent_type.c,
	  njd_set_accent_type/njd_set_accent_type.h,
	  njd_set_accent_type/njd_set_accent_type_rule_ascii_for_euc_jp.h,
	  njd_set_accent_type/njd_set_accent_type_rule_ascii_for_shift_jis.h,
	  njd_set_accent_type/njd_set_accent_type_rule_ascii_for_utf_8.h,
	  njd_set_accent_type/njd_set_accent_type_rule_euc_jp.h,
	  njd_set_accent_type/njd_set_accent_type_rule_shift_jis.h,
	  njd_set_accent_type/njd_set_accent_type_rule_utf_8.h,
	  njd_set_digit/njd_set_digit.c, njd_set_digit/njd_set_digit.h,
	  njd_set_digit/njd_set_digit_rule_ascii_for_euc_jp.h,
	  njd_set_digit/njd_set_digit_rule_ascii_for_shift_jis.h,
	  njd_set_digit/njd_set_digit_rule_ascii_for_utf_8.h,
	  njd_set_digit/njd_set_digit_rule_euc_jp.h,
	  njd_set_digit/njd_set_digit_rule_shift_jis.h,
	  njd_set_digit/njd_set_digit_rule_utf_8.h,
	  njd_set_long_vowel/njd_set_long_vowel.c,
	  njd_set_long_vowel/njd_set_long_vowel.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_ascii_for_euc_jp.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_ascii_for_shift_jis.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_ascii_for_utf_8.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_euc_jp.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_shift_jis.h,
	  njd_set_long_vowel/njd_set_long_vowel_rule_utf_8.h,
	  njd_set_pronunciation/njd_set_pronunciation.c,
	  njd_set_pronunciation/njd_set_pronunciation.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_ascii_for_euc_jp.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_ascii_for_shift_jis.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_ascii_for_utf_8.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_euc_jp.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_shift_jis.h,
	  njd_set_pronunciation/njd_set_pronunciation_rule_utf_8.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel.c,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_ascii_for_euc_jp.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_ascii_for_shift_jis.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_ascii_for_utf_8.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_euc_jp.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_shift_jis.h,
	  njd_set_unvoiced_vowel/njd_set_unvoiced_vowel_rule_utf_8.h,
	  text2mecab/text2mecab.c, text2mecab/text2mecab.h,
	  text2mecab/text2mecab_rule_ascii_for_euc_jp.h,

lib/libopen_jtalk.c
/* libopen_jtalk.c    
 * based on ../bin/open_jtalk.c
 * by Takuya Nishimoto
 * since 2010-06-27    
 */

#define VOICE "/usr/local/share/open_jtalk/hts_voice_nitech_jp_atr503_m001-1.01"
#define DIC   "/usr/local/share/open_jtalk/open_jtalk_dic_utf_8-1.00"

/* ----------------------------------------------------------------- */
/*           The HMM-Based Speech Synthesis System (HTS)             */
/*           Open JTalk developed by HTS Working Group               */
/*           http://open-jtalk.sourceforge.net/                      */
/* ----------------------------------------------------------------- */
/*                                                                   */
/*  Copyright (c) 2008-2010  Nagoya Institute of Technology          */
/*                           Department of Computer Science          */
/*                                                                   */
/* All rights reserved.                                              */
/*                                                                   */
/* Redistribution and use in source and binary forms, with or        */
/* without modification, are permitted provided that the following   */
/* conditions are met:                                               */
/*                                                                   */
/* - Redistributions of source code must retain the above copyright  */
/*   notice, this list of conditions and the following disclaimer.   */
/* - Redistributions in binary form must reproduce the above         */
/*   copyright notice, this list of conditions and the following     */
/*   disclaimer in the documentation and/or other materials provided */
/*   with the distribution.                                          */
/* - Neither the name of the HTS working group nor the names of its  */
/*   contributors may be used to endorse or promote products derived */
/*   from this software without specific prior written permission.   */
/*                                                                   */
/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND            */
/* CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,       */
/* INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF          */
/* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE          */
/* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS */
/* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,          */
/* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED   */
/* TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,     */
/* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON */
/* ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,   */
/* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY    */
/* OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE           */
/* POSSIBILITY OF SUCH DAMAGE.                                       */
/* ----------------------------------------------------------------- */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <math.h>

/* Main headers */
#include "mecab.h"
#include "njd.h"
#include "jpcommon.h"
#include "HTS_engine.h"

/* Sub headers */
#include "text2mecab.h"
#include "mecab2njd.h"
#include "njd_set_pronunciation.h"
#include "njd_set_digit.h"
#include "njd_set_accent_phrase.h"
#include "njd_set_accent_type.h"
#include "njd_set_unvoiced_vowel.h"
#include "njd_set_long_vowel.h"
#include "njd2jpcommon.h"

#define MAXBUFLEN 1024

typedef struct _OpenJTalk {
   Mecab mecab;
   NJD njd;
   JPCommon jpcommon;
   HTS_Engine engine;
} OpenJTalk;

void OpenJTalk_initialize(OpenJTalk * open_jtalk, int sampling_rate, int fperiod, double alpha,
                          int stage, double beta, int audio_buff_size, double uv_threshold,
                          HTS_Boolean use_log_gain, double gv_weight_mcp, double gv_weight_lf0)
{
   Mecab_initialize(&open_jtalk->mecab);
   NJD_initialize(&open_jtalk->njd);
   JPCommon_initialize(&open_jtalk->jpcommon);
   HTS_Engine_initialize(&open_jtalk->engine, 2);
   HTS_Engine_set_sampling_rate(&open_jtalk->engine, sampling_rate);
   HTS_Engine_set_fperiod(&open_jtalk->engine, fperiod);
   HTS_Engine_set_alpha(&open_jtalk->engine, alpha);
   HTS_Engine_set_gamma(&open_jtalk->engine, stage);
   HTS_Engine_set_log_gain(&open_jtalk->engine, use_log_gain);
   HTS_Engine_set_beta(&open_jtalk->engine, beta);
   HTS_Engine_set_audio_buff_size(&open_jtalk->engine, audio_buff_size);
   HTS_Engine_set_msd_threshold(&open_jtalk->engine, 1, uv_threshold);
   HTS_Engine_set_gv_weight(&open_jtalk->engine, 0, gv_weight_mcp);
   HTS_Engine_set_gv_weight(&open_jtalk->engine, 1, gv_weight_lf0);
}

void OpenJTalk_load(OpenJTalk * open_jtalk, char *dn_mecab, char *fn_ms_dur, char *fn_ts_dur,
                    char *fn_ms_mcp, char *fn_ts_mcp, char **fn_ws_mcp, int num_ws_mcp,
                    char *fn_ms_lf0, char *fn_ts_lf0, char **fn_ws_lf0, int num_ws_lf0,
                    char *fn_ms_gvm, char *fn_ts_gvm, char *fn_ms_gvl, char *fn_ts_gvl,
                    char *fn_gv_switch)
{
   Mecab_load(&open_jtalk->mecab, dn_mecab);
   HTS_Engine_load_duration_from_fn(&open_jtalk->engine, &fn_ms_dur, &fn_ts_dur, 1);
   HTS_Engine_load_parameter_from_fn(&open_jtalk->engine, &fn_ms_mcp, &fn_ts_mcp,
                                     fn_ws_mcp, 0, FALSE, num_ws_mcp, 1);
   HTS_Engine_load_parameter_from_fn(&open_jtalk->engine, &fn_ms_lf0, &fn_ts_lf0,
                                     fn_ws_lf0, 1, TRUE, num_ws_lf0, 1);
   if (fn_ms_gvm != NULL) {
      if (fn_ts_gvm != NULL)
         HTS_Engine_load_gv_from_fn(&open_jtalk->engine, &fn_ms_gvm, &fn_ts_gvm, 0, 1);
      else
         HTS_Engine_load_gv_from_fn(&open_jtalk->engine, &fn_ms_gvm, NULL, 0, 1);
   }
   if (fn_ms_gvl != NULL) {
      if (fn_ts_gvl != NULL)
         HTS_Engine_load_gv_from_fn(&open_jtalk->engine, &fn_ms_gvl, &fn_ts_gvl, 1, 1);
      else
         HTS_Engine_load_gv_from_fn(&open_jtalk->engine, &fn_ms_gvl, NULL, 1, 1);
   }
   if (fn_gv_switch != NULL)
      HTS_Engine_load_gv_switch_from_fn(&open_jtalk->engine, fn_gv_switch);
}

void OpenJTalk_synthesis(OpenJTalk * open_jtalk, char *txt, FILE * wavfp, FILE * logfp)
{
   char *buff = (char *) calloc(2 * strlen(txt) + 1, sizeof(char));

   text2mecab(buff, txt);
   Mecab_analysis(&open_jtalk->mecab, buff);
   free(buff);
   mecab2njd(&open_jtalk->njd, Mecab_get_feature(&open_jtalk->mecab),
             Mecab_get_size(&open_jtalk->mecab));
   njd_set_pronunciation(&open_jtalk->njd);
   njd_set_digit(&open_jtalk->njd);
   njd_set_accent_phrase(&open_jtalk->njd);
   njd_set_accent_type(&open_jtalk->njd);
   njd_set_unvoiced_vowel(&open_jtalk->njd);
   njd_set_long_vowel(&open_jtalk->njd);
   njd2jpcommon(&open_jtalk->jpcommon, &open_jtalk->njd);
   JPCommon_make_label(&open_jtalk->jpcommon);
   if (JPCommon_get_label_size(&open_jtalk->jpcommon) > 2) {
      HTS_Engine_load_label_from_string_list(&open_jtalk->engine,
                                             JPCommon_get_label_feature(&open_jtalk->jpcommon),
                                             JPCommon_get_label_size(&open_jtalk->jpcommon));
      HTS_Engine_create_sstream(&open_jtalk->engine);
      HTS_Engine_create_pstream(&open_jtalk->engine);
      HTS_Engine_create_gstream(&open_jtalk->engine);
#if 0
      if (wavfp != NULL)
         HTS_Engine_save_riff(&open_jtalk->engine, wavfp);
#else
      /* based on HTS_engine_API:
       * HTS_Engine_save_generated_speech: output generated speech */
      {
        HTS_Engine *engine = &open_jtalk->engine;
        int i;
        short temp;
        HTS_GStreamSet *gss = &engine->gss;
        for (i = 0; i < HTS_GStreamSet_get_total_nsample(gss); i++) {
          temp = HTS_GStreamSet_get_speech(gss, i);
          fwrite(&temp, sizeof(short), 1, wavfp);
        }
      }
#endif
#if 0
      if (logfp != NULL) {
         fprintf(logfp, "[Text analysis result]\n");
         NJD_fprint(&open_jtalk->njd, logfp);
         fprintf(logfp, "\n[Output label]\n");
         HTS_Engine_save_label(&open_jtalk->engine, logfp);
         fprintf(logfp, "\n");
         HTS_Engine_save_information(&open_jtalk->engine, logfp);
      }
#endif
      HTS_Engine_refresh(&open_jtalk->engine);
   }
   JPCommon_refresh(&open_jtalk->jpcommon);
   NJD_refresh(&open_jtalk->njd);
   Mecab_refresh(&open_jtalk->mecab);
}

void OpenJTalk_clear(OpenJTalk * open_jtalk)
{
   Mecab_clear(&open_jtalk->mecab);
   NJD_clear(&open_jtalk->njd);
   JPCommon_clear(&open_jtalk->jpcommon);
   HTS_Engine_clear(&open_jtalk->engine);
}

void Usage()
{
   fprintf(stderr, "\n");
   fprintf(stderr, "The HMM-based speech synthesis system (HTS)\n");
   fprintf(stderr, "Open JTalk version 1.01 (http://open-jtalk.sourceforge.net/)\n");
   fprintf(stderr, "Copyright (C) 2008-2010  Nagoya Institute of Technology\n");
   fprintf(stderr, "All rights reserved.\n");
   HTS_show_copyright(stderr);
   fprintf(stderr, "\n");
   fprintf(stderr, "Yet Another Part-of-Speech and Morphological Analyzer (Mecab)\n");
   fprintf(stderr, "mecab version 0.98 (http://mecab.sourceforge.net/)\n");
   fprintf(stderr, "Copyright (C) 2001-2008  Taku Kudo\n");
   fprintf(stderr, "              2004-2008  Nippon Telegraph and Telephone Corporation\n");
   fprintf(stderr, "All rights reserved.\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "NAIST Japanese dictionary\n");
   fprintf(stderr, "mecab-naist-jdic version 0.6.1-20090630 (http://naist-jdic.sourceforge.jp/)\n");
   fprintf(stderr, "Copyright (C) 2009  Nara Institute of Science and Technology\n");
   fprintf(stderr, "All rights reserved.\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "open_jtalk - An HMM-based text to speech system\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "  usage:\n");
   fprintf(stderr, "       open_jtalk [ options ] [ infile ] \n");
   fprintf(stderr,
           "  options:                                                                   [  def][ min--max]\n");
   fprintf(stderr,
           "    -x dir         : dictionary directory                                    [  N/A]\n");
   fprintf(stderr,
           "    -td tree       : decision trees file for state duration                  [  N/A]\n");
   fprintf(stderr,
           "    -tf tree       : decision trees file for Log F0                          [  N/A]\n");
   fprintf(stderr,
           "    -tm tree       : decision trees file for spectrum                        [  N/A]\n");
   fprintf(stderr,
           "    -md pdf        : model file for state duration                           [  N/A]\n");
   fprintf(stderr,
           "    -mf pdf        : model file for Log F0                                   [  N/A]\n");
   fprintf(stderr,
           "    -mm pdf        : model file for spectrum                                 [  N/A]\n");
   fprintf(stderr,
           "    -df win        : window files for calculation delta of Log F0            [  N/A]\n");
   fprintf(stderr,
           "    -dm win        : window files for calculation delta of spectrum          [  N/A]\n");
   fprintf(stderr,
           "    -ow s          : filename of output wav audio (generated speech)         [  N/A]\n");
   fprintf(stderr,
           "    -ot s          : filename of output trace information                    [  N/A]\n");
   fprintf(stderr,
           "    -s  i          : sampling frequency                                      [16000][   1--48000]\n");
   fprintf(stderr,
           "    -p  i          : frame period (point)                                    [   80][   1--]\n");
   fprintf(stderr,
           "    -a  f          : all-pass constant                                       [ 0.42][ 0.0--1.0]\n");
   fprintf(stderr,
           "    -g  i          : gamma = -1 / i (if i=0 then gamma=0)                    [    0][   0-- ]\n");
   fprintf(stderr,
           "    -b  f          : postfiltering coefficient                               [  0.0][-0.8--0.8]\n");
   fprintf(stderr,
           "    -l             : regard input as log gain and output linear one (LSP)    [  N/A]\n");
   fprintf(stderr,
           "    -u  f          : voiced/unvoiced threshold                               [  0.5][ 0.0--1.0]\n");
   fprintf(stderr,
           "    -ef tree       : decision tree file for GV of Log F0                     [  N/A]\n");
   fprintf(stderr,
           "    -em tree       : decision tree file for GV of spectrum                   [  N/A]\n");
   fprintf(stderr,
           "    -cf pdf        : filename of GV for Log F0                               [  N/A]\n");
   fprintf(stderr,
           "    -cm pdf        : filename of GV for spectrum                             [  N/A]\n");
   fprintf(stderr,
           "    -jf f          : weight of GV for Log F0                                 [  0.7][ 0.0--2.0]\n");
   fprintf(stderr,
           "    -jm f          : weight of GV for spectrum                               [  1.0][ 0.0--2.0]\n");
   fprintf(stderr,
           "    -k  tree       : use GV switch                                           [  N/A]\n");
   fprintf(stderr,
           "    -z  i          : audio buffer size                                       [ 1600][   0--48000]\n");
   fprintf(stderr, "  infile:\n");
   fprintf(stderr,
           "    text file                                                                [stdin]\n");
   fprintf(stderr, "  note:\n");
   fprintf(stderr, "    option '-d' may be repeated to use multiple delta parameters.\n");
   fprintf(stderr, "    generated spectrum and log F0 sequences are saved in natural\n");
   fprintf(stderr, "    endian, binary (float) format.\n");
   fprintf(stderr, "\n");

   exit(0);
}

/* Getfp: wrapper for fopen */
FILE *Getfp(const char *name, const char *opt)
{
   FILE *fp = fopen(name, opt);

   if (fp == NULL) {
      fprintf(stderr, "ERROR: Getfp() in open_jtalk.c: Cannot open %s.\n", name);
      exit(1);
   }

   return (fp);
}

int libopen_jtalk_main(int argc, char **argv)
{
   FILE *txtfp = stdin;
   char *txtfn = NULL;
   FILE *wavfp = NULL;
   FILE *logfp = NULL;

   /* text */
   char buff[MAXBUFLEN];

   /* engine */
   OpenJTalk open_jtalk;

   /* directory name of dictionary */
   char *dn_mecab = NULL;

   /* file names of models */
   char *fn_ms_lf0 = NULL;
   char *fn_ms_mcp = NULL;
   char *fn_ms_dur = NULL;

   /* file names of trees */
   char *fn_ts_lf0 = NULL;
   char *fn_ts_mcp = NULL;
   char *fn_ts_dur = NULL;

   /* file names of windows */
   char **fn_ws_lf0;
   char **fn_ws_mcp;
   int num_ws_lf0 = 0, num_ws_mcp = 0;

   /* file names of global variance */
   char *fn_ms_gvl = NULL;
   char *fn_ms_gvm = NULL;

   /* file names of global variance trees */
   char *fn_ts_gvl = NULL;
   char *fn_ts_gvm = NULL;

   /* file names of global variance switch */
   char *fn_gv_switch = NULL;

   /* global parameter */
   int sampling_rate = 16000;
   int fperiod = 80;
   double alpha = 0.42;
   int stage = 0;               /* gamma = -1.0/stage */
   double beta = 0.0;
   int audio_buff_size = 1600;
   double uv_threshold = 0.5;
   double gv_weight_lf0 = 0.7;
   double gv_weight_mcp = 1.0;
   HTS_Boolean use_log_gain = FALSE;

   /* parse command line */
   if (argc == 1)
      Usage();

   /* delta window handler for log f0 */
   fn_ws_lf0 = (char **) calloc(argc, sizeof(char *));
   /* delta window handler for mel-cepstrum */
   fn_ws_mcp = (char **) calloc(argc, sizeof(char *));

   /* read command */
   while (--argc) {
      if (**++argv == '-') {
         switch (*(*argv + 1)) {
         case 'x':
            dn_mecab = *(++argv);
            --argc;
            break;
         case 't':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               fn_ts_lf0 = *(++argv); /* -tf $VOICE/tree-lf0.inf */
               break;
            case 'm':
               fn_ts_mcp = *(++argv); /* -tm $VOICE/tree-mgc.inf */
               break;
            case 'd':
               fn_ts_dur = *(++argv); /* -td $VOICE/tree-dur.inf */
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-t%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'm':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               fn_ms_lf0 = *(++argv); /* -mf $VOICE/lf0.pdf */
               break;
            case 'm':
               fn_ms_mcp = *(++argv); /* -mm $VOICE/mgc.pdf */
               break;
            case 'd':
               fn_ms_dur = *(++argv); /* -md $VOICE/dur.pdf */
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-m%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'd':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               fn_ws_lf0[num_ws_lf0] = *(++argv); /* -df $VOICE/lf0.win1 -df $VOICE/lf0.win2 -df $VOICE/lf0.win3 */
               num_ws_lf0++;
               break;
            case 'm':
               fn_ws_mcp[num_ws_mcp] = *(++argv); /* -dm $VOICE/mgc.win1 -dm $VOICE/mgc.win2 -dm $VOICE/mgc.win3 */
               num_ws_mcp++;
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-d%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'o':
            switch (*(*argv + 2)) {
            case 'w':
               wavfp = Getfp(*(++argv), "wb"); /* -ow _hoge.wav */
               break;
            case 't':
               logfp = Getfp(*(++argv), "w"); /* -ot _log.txt */
               break;
            default:
               fprintf(stderr, "ERROR: main() in open_jtalk.c: Invalid option '-o%c'.\n",
                       *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'h':
            Usage();
            break;
         case 's':
            sampling_rate = atoi(*++argv);
            --argc;
            break;
         case 'p':
            fperiod = atoi(*++argv);
            --argc;
            break;
         case 'a':
            alpha = atof(*++argv);
            --argc;
            break;
         case 'g':
            stage = atoi(*++argv);
            --argc;
            break;
         case 'l':
            use_log_gain = TRUE;
            break;
         case 'b':
            beta = atof(*++argv);
            --argc;
            break;
         case 'u':
            uv_threshold = atof(*++argv);
            --argc;
            break;
         case 'e':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               fn_ts_gvl = *(++argv); /* -ef $VOICE/tree-gv-lf0.inf */
               break;
            case 'm':
               fn_ts_gvm = *(++argv); /* -em $VOICE/tree-gv-mgc.inf */
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-e%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'c':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               fn_ms_gvl = *(++argv); /* -cf $VOICE/gv-lf0.pdf */
               break;
            case 'm':
               fn_ms_gvm = *(++argv); /* -cm $VOICE/gv-mgc.pdf */
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-c%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'j':
            switch (*(*argv + 2)) {
            case 'f':
            case 'p':
               gv_weight_lf0 = atof(*(++argv));
               break;
            case 'm':
               gv_weight_mcp = atof(*(++argv));
               break;
            default:
               fprintf(stderr,
                       "ERROR: main() in open_jtalk.c: Invalid option '-j%c'.\n", *(*argv + 2));
               exit(1);
            }
            --argc;
            break;
         case 'k':
            fn_gv_switch = *++argv; /* -k $VOICE/gv-switch.inf */
            --argc;
            break;
         case 'z':
            audio_buff_size = atoi(*++argv);
            --argc;
            break;
         default:
            fprintf(stderr, "ERROR: main() in open_jtalk.c: Invalid option '-%c'.\n", *(*argv + 1));
            exit(1);
         }
      } else {
         txtfn = *argv;
         txtfp = Getfp(txtfn, "rt");
      }
   }
#if 0
   /* dictionary directory check */
   if (dn_mecab == NULL) {
      fprintf(stderr, "ERROR: main() in open_jtalk.c: No dictionary directory.\n");
      exit(1);
   }
   /* number of models,trees check */
   if (fn_ms_dur == NULL || fn_ms_mcp == NULL || fn_ms_lf0 == NULL ||
       fn_ts_dur == NULL || fn_ts_mcp == NULL || fn_ts_lf0 == NULL ||
       fn_ws_mcp == NULL || fn_ws_lf0 == NULL) {
      fprintf(stderr,
              "ERROR: main() in open_jtalk.c: Specify models (trees) for each parameter.\n");
      exit(1);
   }
#endif
   /* initialize and load */
   OpenJTalk_initialize(&open_jtalk, sampling_rate, fperiod, alpha, stage, beta, audio_buff_size,
                        uv_threshold, use_log_gain, gv_weight_mcp, gv_weight_lf0);
   fn_ws_mcp[0] = VOICE "/mgc.win1";
   fn_ws_mcp[1] = VOICE "/mgc.win2";
   fn_ws_mcp[2] = VOICE "/mgc.win3";
   fn_ws_lf0[0] = VOICE "/lf0.win1";
   fn_ws_lf0[1] = VOICE "/lf0.win2";
   fn_ws_lf0[2] = VOICE "/lf0.win3";
   OpenJTalk_load(&open_jtalk,
		  DIC, /* dn_mecab */ 
		  VOICE "/dur.pdf", /* fn_ms_dur */
		  VOICE "/tree-dur.inf", /* fn_ts_dur */
		  VOICE "/mgc.pdf", /* fn_ms_mcp */
		  VOICE "/tree-mgc.inf", /* fn_ts_mcp */
		  fn_ws_mcp,
		  3, /* num_ws_mcp */
		  VOICE "/lf0.pdf", /* fn_ms_lf0 */
		  VOICE "/tree-lf0.inf", /* fn_ts_lf0 */
		  fn_ws_lf0, 
		  3, /* num_ws_lf0 */
		  VOICE "/gv-mgc.pdf", /* fn_ms_gvm */
		  VOICE "/tree-gv-mgc.inf", /* fn_ts_gvm */
		  VOICE "/gv-lf0.pdf",  /* fn_ms_gvl */
		  VOICE "/tree-gv-lf0.inf", /* fn_ts_gvl */
		  VOICE "/gv-switch.inf" /* fn_gv_switch */
		  );

   /* synthesis */
   fgets(buff, MAXBUFLEN - 1, txtfp);
   OpenJTalk_synthesis(&open_jtalk, buff, wavfp, logfp);

   /* free */
   OpenJTalk_clear(&open_jtalk);
   free(fn_ws_mcp);
   free(fn_ws_lf0);
   if (txtfn != NULL)
      fclose(txtfp);
   if (wavfp != NULL)
      fclose(wavfp);
   if (logfp != NULL)
      fclose(logfp);

   return 0;
}
